File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    int[][] dp;
    
    private int isPalindrome(String s, int i, int j) {
        if (i >= j) {
            return 1;
        }
        
        if (dp[i][j] != -1) return dp[i][j];
        
        if (s.charAt(i) == s.charAt(j)) {
            return dp[i][j] = isPalindrome(s, i + 1, j - 1);
        }
        
        return 0;
    }
    
    public String longestPalindrome(String s) {
        int n = s.length(), maxLen = 0, startPos = 0;
        dp = new int[n + 1][n + 1];
        for (int[] cur : dp) {
            Arrays.fill(cur, -1);
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (isPalindrome(s, i, j) == 1) {
                    if (j - i + 1 > maxLen) {
                        maxLen = j - i + 1;
                        startPos = i;
                    }
                }
            }
        }
        
        return s.substring(startPos, startPos + maxLen);
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string f(int c1, int c2, string& s, int& n) {
        while (c1 >= 0 && c2 < n && s[c1] == s[c2]) {
            c1--;
            c2++;
        }
        return s.substr(c1 + 1, c2 - c1 - 1);
    }
    
    string longestPalindrome(string s) {
        if (s.size() <= 1) {
            return s;
        }
        
        string res = "";
        
        for (int i = 0, n = s.size(); i < n - 1; i++) {
            string s1 = f(i, i, s, n);
            if (s1.size() > res.size()) {
                res = s1;
            }
            s1 = f(i, i + 1, s, n);
            if (s1.size() > res.size()) {
                res = s1;
            }
        }
        
        return res;
    }
};